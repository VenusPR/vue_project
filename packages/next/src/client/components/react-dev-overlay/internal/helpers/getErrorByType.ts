import {
  ACTION_UNHANDLED_ERROR,
  ACTION_UNHANDLED_REJECTION,
} from '../error-overlay-reducer'
import { SupportedErrorEvent } from '../container/Errors'
import { getErrorSource } from './nodeStackFrames'
import { getOriginalStackFrames, OriginalStackFrame } from './stack-frame'
import { ComponentStackFrame } from './parse-component-stack'

export type ReadyRuntimeError = {
  id: number
  runtime: true
  error: Error
  frames: OriginalStackFrame[]
  componentStackFrames?: ComponentStackFrame[]
}

export async function getErrorByType(
  ev: SupportedErrorEvent
): Promise<ReadyRuntimeError> {
  const { id, event } = ev
  switch (event.type) {
    case ACTION_UNHANDLED_ERROR:
    case ACTION_UNHANDLED_REJECTION: {
      const readyRuntimeError: ReadyRuntimeError = {
        id,
        runtime: true,
        error: event.reason,
        frames: await getOriginalStackFrames(
          event.frames,
          getErrorSource(event.reason),
          event.reason.toString()
        ),
      }
      if (event.type === ACTION_UNHANDLED_ERROR) {
        readyRuntimeError.componentStackFrames = event.componentStackFrames
      }
      return readyRuntimeError
    }
    default: {
      break
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const _: never = event
  throw new Error('type system invariant violation')
}
