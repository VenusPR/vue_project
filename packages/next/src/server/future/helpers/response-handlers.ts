export function handleTemporaryRedirectResponse(url: string): Response {
  return new Response(null, {
    status: 302,
    statusText: 'Found',
    headers: {
      location: url,
    },
  })
}

export function handleBadRequestResponse(): Response {
  return new Response(null, {
    status: 400,
    statusText: 'Bad Request',
  })
}

export function handleNotFoundResponse(): Response {
  return new Response(null, {
    status: 404,
    statusText: 'Not Found',
  })
}

export function handleMethodNotAllowedResponse(): Response {
  return new Response(null, {
    status: 405,
    statusText: 'Method Not Allowed',
  })
}

export function handleInternalServerErrorResponse(): Response {
  return new Response(null, {
    status: 500,
    statusText: 'Internal Server Error',
  })
}
