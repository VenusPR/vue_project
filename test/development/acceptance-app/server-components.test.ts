/* eslint-env jest */
import { sandbox } from './helpers'
import { createNext, FileRef } from 'e2e-utils'
import { NextInstance } from 'test/lib/next-modes/base'
import path from 'path'

describe('Error Overlay for server components', () => {
  if (process.env.NEXT_TEST_REACT_VERSION === '^17') {
    it('should skip for react v17', () => {})
    return
  }

  let next: NextInstance

  beforeAll(async () => {
    next = await createNext({
      files: new FileRef(path.join(__dirname, 'fixtures', 'default-template')),
      dependencies: {
        react: 'latest',
        'react-dom': 'latest',
      },
      skipStart: true,
    })
  })
  afterAll(() => next.destroy())

  describe('createContext called in Server Component', () => {
    it('should show error when React.createContext is called', async () => {
      const { session, browser, cleanup } = await sandbox(
        next,
        new Map([
          [
            'app/page.js',
            `
        import React from 'react'
        const Context = React.createContext()
        export default function Page() {
          return (
            <>
                <Context.Provider value="hello">
                    <h1>Page</h1>
                </Context.Provider>
            </>
          )
        }`,
          ],
        ])
      )

      // TODO-APP: currently requires a full reload because moving from a client component to a server component isn't causing a Fast Refresh yet.
      await browser.refresh()

      expect(await session.hasRedbox(true)).toBe(true)
      expect(await session.getRedboxSource(true)).toMatchSnapshot()
      expect(next.cliOutput).toContain(
        'createContext only works in Client Components'
      )

      await cleanup()
    })

    it('should show error when React.createContext is called in external package', async () => {
      const { session, browser, cleanup } = await sandbox(
        next,
        new Map([
          [
            'node_modules/my-package/index.js',
            `
          const React = require('react')
          module.exports = React.createContext()
        `,
          ],
          [
            'node_modules/my-package/package.json',
            `
          {
            "name": "my-package",
            "version": "0.0.1"
          }
        `,
          ],
          [
            'app/page.js',
            `
        import Context from 'my-package'
        export default function Page() {
          return (
            <>
                <Context.Provider value="hello">
                    <h1>Page</h1>
                </Context.Provider>
            </>
          )
        }`,
          ],
        ])
      )

      // TODO-APP: currently requires a full reload because moving from a client component to a server component isn't causing a Fast Refresh yet.
      await browser.refresh()

      expect(await session.hasRedbox(true)).toBe(true)
      expect(await session.getRedboxSource(true)).toMatchSnapshot()
      expect(next.cliOutput).toContain(
        'createContext only works in Client Components'
      )

      await cleanup()
    })

    it('should show error when createContext is called in external package', async () => {
      const { session, browser, cleanup } = await sandbox(
        next,
        new Map([
          [
            'node_modules/my-package/index.js',
            `
          const { createContext } = require('react')
          module.exports = createContext()
        `,
          ],
          [
            'node_modules/my-package/package.json',
            `
          {
            "name": "my-package",
            "version": "0.0.1"
          }
        `,
          ],
          [
            'app/page.js',
            `
        import Context from 'my-package'
        export default function Page() {
          return (
            <>
                <Context.Provider value="hello">
                    <h1>Page</h1>
                </Context.Provider>
            </>
          )
        }`,
          ],
        ])
      )

      // TODO-APP: currently requires a full reload because moving from a client component to a server component isn't causing a Fast Refresh yet.
      await browser.refresh()

      expect(await session.hasRedbox(true)).toBe(true)
      expect(await session.getRedboxSource(true)).toMatchSnapshot()
      expect(next.cliOutput).toContain(
        'createContext only works in Client Components'
      )

      await cleanup()
    })
  })
})
