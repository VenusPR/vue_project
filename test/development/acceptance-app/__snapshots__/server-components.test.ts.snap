// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Overlay for server components createContext called in Server Component should show error when React.createContext is called 1`] = `
" 1 of 1 unhandled error
Server Error

TypeError: createContext only works in Client Components. Add the \\"use client\\" directive at the top of the file to use it. Read more: https://beta.nextjs.org/docs/rendering/server-and-client-components#context

This error happened while generating the page. Any console logs will be displayed in the terminal window.

app/page.js (3:24) @ React

  1 | 
  2 | import React from 'react'
> 3 | const Context = React.createContext()
    |                ^
  4 | export default function Page() {
  5 |   return (
  6 |     <>"
`;

exports[`Error Overlay for server components createContext called in Server Component should show error when React.createContext is called in external package 1`] = `
" 1 of 1 unhandled error
Server Error

TypeError: createContext only works in Client Components. Add the \\"use client\\" directive at the top of the file to use it. Read more: https://beta.nextjs.org/docs/rendering/server-and-client-components#context

This error happened while generating the page. Any console logs will be displayed in the terminal window.

null"
`;

exports[`Error Overlay for server components createContext called in Server Component should show error when createContext is called in external package 1`] = `
" 1 of 1 unhandled error
Server Error

TypeError: createContext only works in Client Components. Add the \\"use client\\" directive at the top of the file to use it. Read more: https://beta.nextjs.org/docs/rendering/server-and-client-components#context

This error happened while generating the page. Any console logs will be displayed in the terminal window.

null"
`;
